File: cs/FaceSentimentAnalyzer/FaceSentimentAnalyzerDescriptor.cs
// Copyright (c) Microsoft Corporation. All rights reserved. 

using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading.Tasks;
using Microsoft.AI.Skills.SkillInterfacePreview;
using Windows.Foundation;
using Windows.Graphics.Imaging;

namespace FaceSentimentAnalyzer
{
    /// <summary>
    /// FaceSentimentAnalyzerDescriptor class
    /// Exposes information about the skill and its input/output variable requirements.
    /// Also acts as a factory for FaceSentimentAnalyzerSkill
    /// </summary>
    public sealed class FaceSentimentAnalyzerDescriptor : ISkillDescriptor
    {
        // Member variables
        private List<ISkillFeatureDescriptor> m_inputSkillDesc;
        private List<ISkillFeatureDescriptor> m_outputSkillDesc;

        /// <summary>
        /// FaceSentimentAnalyzerDescriptor constructor
        /// </summary>
        public FaceSentimentAnalyzerDescriptor()
        {
            Name = "FaceSentimentAnalyzer";

            Description = "Finds a face in the image and infers its predominant sentiment from a set of 8 possible labels";
           
            // {F8D275CE-C244-4E71-8A39-57335D291388}
            Id = new Guid(0xf8d275ce, 0xc244, 0x4e71, 0x8a, 0x39, 0x57, 0x33, 0x5d, 0x29, 0x13, 0x88);

            Version = SkillVersion.Create(
                0,  // major version
                1,  // minor version
                "Contoso Developer", // Author name 
                "Contoso Publishing" // Publisher name
                ); 

            // Describe input feature
            m_inputSkillDesc = new List<ISkillFeatureDescriptor>();
            m_inputSkillDesc.Add(
                SkillFeatureImageDescriptor.Create(
                    FaceSentimentAnalyzerConst.SKILL_INPUTNAME_IMAGE,
                    "the input image onto which the sentiment analysis runs",
                    true, // isRequired (since this is an input, it is required to be bound before the evaluation occurs)
                    -1, // width
                    -1, // height
                    -1, // maxDimension
                    BitmapPixelFormat.Nv12,
                    BitmapAlphaMode.Ignore)
            );

            // Describe first output feature
            m_outputSkillDesc = new List<ISkillFeatureDescriptor>();
            m_outputSkillDesc.Add(
                SkillFeatureTensorDescriptor.Create(
                    FaceSentimentAnalyzerConst.SKILL_OUTPUTNAME_FACERECTANGLE,
                    "a face bounding box in relative coordinates (left, top, right, bottom)",
                    false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                    new List<long>() { 4 }, // tensor shape
                    SkillElementKind.Float)
                );

            // Describe second output feature
            m_outputSkillDesc.Add(
                SkillFeatureTensorDescriptor.Create(
                    FaceSentimentAnalyzerConst.SKILL_OUTPUTNAME_FACESENTIMENTSCORES,
                    "the prediction score for each class",
                    false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                    new List<long>() { 1, 8 }, // tensor shape
                    SkillElementKind.Float)
                );
        }

        /// <summary>
        /// Retrieves a list of supported ISkillExecutionDevice to run the skill logic on
        /// </summary>
        /// <returns></returns>
        public IAsyncOperation<IReadOnlyList<ISkillExecutionDevice>> GetSupportedExecutionDevicesAsync()
        {
            return AsyncInfo.Run(async (token) =>
            {
                var result = new List<ISkillExecutionDevice>();
                await Task.Run(() =>
                {
                    // Add CPU as supported device
                    result.Add(SkillExecutionDeviceCPU.Create());

                    // Retrieve a list of GPUs available on the system and filter them bv keeping only GPUs that support DX11+ feature level
                    var gpuDevices = SkillExecutionDeviceGPU.GetAvailableGpuExecutionDevices();
                    var compatibleGpuDevices = gpuDevices.Where((device) => (device as SkillExecutionDeviceGPU).MaxSupportedFeatureLevel >= D3DFeatureLevelKind.D3D_FEATURE_LEVEL_11_0);
                    result.AddRange(compatibleGpuDevices);
                });
                return result as IReadOnlyList<ISkillExecutionDevice>;
            });
        }

        /// <summary>
        /// Factory method for instantiating and initializing the skill
        /// </summary>
        /// <param name="executionDevice"></param>
        /// <returns></returns>
        public IAsyncOperation<ISkill> CreateSkillAsync(ISkillExecutionDevice executionDevice)
        {
            return AsyncInfo.Run(async (token) =>
            {
                // Create a skill instance with the executionDevice supplied
                var skillInstance = await FaceSentimentAnalyzerSkill.CreateAsync(this, executionDevice);          

                return skillInstance as ISkill;
            });
        }

        /// <summary>
        /// Returns a description of the skill
        /// </summary>
        public string Description { get; private set; }

        /// <summary>
        /// Returns a unique skill identifier
        /// </summary>
        public Guid Id { get; }

        /// <summary>
        /// Returns a list of descriptors that correlate with each input SkillFeature
        /// </summary>
        public IReadOnlyList<ISkillFeatureDescriptor> InputFeatureDescriptors => m_inputSkillDesc;

        /// <summary>
        /// Returns a set of metadata that may control the skill execution differently than by feeding an input 
        /// i.e. internal state, sub-process execution frequency, etc.
        /// </summary>
        public IReadOnlyDictionary<string, string> Metadata => null;

        /// <summary>
        /// Returns the skill name
        /// </summary>
        public string Name { get; private set; }

        /// <summary>
        /// Returns a list of descriptors that correlate with each output SkillFeature
        /// </summary>
        public IReadOnlyList<ISkillFeatureDescriptor> OutputFeatureDescriptors => m_outputSkillDesc;

        /// <summary>
        /// Returns the Version information of the skill
        /// </summary>
        public SkillVersion Version { get; }
    }
}

File: samples/SentimentAnalyzerCustomSkill/cpp/FaceSentimentAnalyzer/FaceSentimentAnalyzerDescriptor.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.

#include "pch.h"
#include "FaceSentimentAnalyzerConst.h"
#include "FaceSentimentAnalyzerDescriptor.h"
#include "FaceSentimentAnalyzerSkill.h"
#include "winrt/Microsoft.AI.Skills.DXCoreExecutionDevice.h"

using namespace winrt::Microsoft::AI::Skills::SkillInterfacePreview;
using namespace winrt::Microsoft::AI::Skills::DXCoreExecutionDevice;

namespace winrt::Contoso::FaceSentimentAnalyzer::implementation
{
    // {678BD455-4190-45D3-B5DA-41543283C092}
    const guid FaceSentimentAnalyzerId = guid(0x678bd455, 0x4190, 0x45d3, { 0xb5, 0xda, 0x41, 0x54, 0x32, 0x83, 0xc0, 0x92 });

    //
    // FaceSentimentAnalyzerDescriptor constructor
    //
    FaceSentimentAnalyzerDescriptor::FaceSentimentAnalyzerDescriptor()
    {
        m_version = SkillVersion::Create(
            0,  // major version
            1,  // minor version
            L"Contoso Developer", // Author name 
            L"Contoso Publishing" // Publisher name
        );

        auto inputSkillDesc = single_threaded_vector<ISkillFeatureDescriptor>();
        auto outputSkillDesc = single_threaded_vector<ISkillFeatureDescriptor>();

        // Describe input feature
        inputSkillDesc.Append(
            SkillFeatureImageDescriptor(
                SKILL_INPUTNAME_IMAGE,
                L"the input image onto which the sentiment analysis runs",
                true, // isRequired (since this is an input, it is required to be bound before the evaluation occurs)
                -1, // width
                -1, // height
                -1, // maxDimension
                Windows::Graphics::Imaging::BitmapPixelFormat::Nv12,
                Windows::Graphics::Imaging::BitmapAlphaMode::Ignore)
        );

        // Describe first output feature
        outputSkillDesc.Append(
            SkillFeatureTensorDescriptor(
                SKILL_OUTPUTNAME_FACERECTANGLE,
                L"a face bounding box in relative coordinates (left, top, right, bottom)",
                false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                single_threaded_vector<int64_t>({ 4 }).GetView(), // tensor shape
                SkillElementKind::Float)
        );

        // Describe second output feature
        outputSkillDesc.Append(
            SkillFeatureTensorDescriptor(
                SKILL_OUTPUTNAME_FACESENTIMENTSCORES,
                L"the prediction score for each class",
                false, // isRequired (since this is an output, it automatically get populated after the evaluation occurs)
                single_threaded_vector<int64_t>({ 1, 8 }).GetView(), // tensor shape
                SkillElementKind::Float)
        );

        m_inputSkillDesc = inputSkillDesc.GetView();
        m_outputSkillDesc = outputSkillDesc.GetView();
    }

    //
    // Retrieves a list of supported ISkillExecutionDevice to run the skill logic on
    //
    Windows::Foundation::IAsyncOperation<Windows::Foundation::Collections::IVectorView<ISkillExecutionDevice>> FaceSentimentAnalyzerDescriptor::GetSupportedExecutionDevicesAsync()
    {
        m_devices = single_threaded_vector<ISkillExecutionDevice>();
        m_devices.Append(SkillExecutionDeviceCPU::Create());
        auto devices = SkillExecutionDeviceDXCore::GetAvailableHardwareExecutionDevices();
        for (auto iter : devices)
        {
            m_devices.Append(iter);
        }
        co_await resume_background();
        return m_devices.GetView();
    }

    //
    // Factory method for instantiating and initializing the skill using the optimal execution device available
    //
    Windows::Foundation::IAsyncOperation<ISkill> FaceSentimentAnalyzerDescriptor::CreateSkillAsync()
    {
        auto supportedDevices = GetSupportedExecutionDevicesAsync().get();
        ISkillExecutionDevice deviceToUse = supportedDevices.First().Current();

        // Use the first GPU or VPU device if available, otherwise default to CPU
        for (auto device : supportedDevices)
        {
            if (device.ExecutionDeviceKind() == SkillExecutionDeviceKind::Gpu
                || device.ExecutionDeviceKind() == SkillExecutionDeviceKind::Vpu)
            {
                deviceToUse = device;
                break;
            }
        }
        return CreateSkillAsync(deviceToUse);
    }

    //
    // Factory method for instantiating and initializing the skill
    //
    Windows::Foundation::IAsyncOperation<ISkill> FaceSentimentAnalyzerDescriptor::CreateSkillAsync(ISkillExecutionDevice const executionDevice)
    {
        co_await resume_background();
        auto desc = this->operator winrt::Windows::Foundation::IInspectable().as<ISkillDescriptor>();
        auto skill = FaceSentimentAnalyzerSkill::CreateAsync(desc, executionDevice).get();
        return skill;
    }

    winrt::guid FaceSentimentAnalyzerDescriptor::Id()
    {
        return FaceSentimentAnalyzerId;
    }

    hstring FaceSentimentAnalyzerDescriptor::Name()
    {
        return L"FaceSentimentAnalyzer";
    }
    
    hstring FaceSentimentAnalyzerDescriptor::Description()
    {
        return L"Finds a face in the image and infers its predominant sentiment from a set of 8 possible labels";
    }

    SkillVersion FaceSentimentAnalyzerDescriptor::Version()
    {
        return m_version;
    }

    Windows::Foundation::Collections::IVectorView<ISkillFeatureDescriptor> FaceSentimentAnalyzerDescriptor::InputFeatureDescriptors()
    {
        return m_inputSkillDesc;
    }

    Windows::Foundation::Collections::IVectorView<ISkillFeatureDescriptor> FaceSentimentAnalyzerDescriptor::OutputFeatureDescriptors()
    {
        return m_outputSkillDesc;
    }

    Windows::Foundation::Collections::IMapView<hstring, hstring> FaceSentimentAnalyzerDescriptor::Metadata()
    {
        return single_threaded_map<hstring, hstring>().GetView();
    }

    
}
